### ISSUE TITLE
MAE results do not match with the paper or am I using it wrong?

### ISSUE THREAD

**Issue by roxanneluo** [facebookresearch/sapiens#108]
Hi,

I really like your work and am interested in testing out the reconstruction capabilities of the pretrained MAE part. To do that, I downloaded the pretrained mae from https://huggingface.co/facebook/sapiens-pretrain-1b/tree/main by
```
wget https://huggingface.co/facebook/sapiens-pretrain-1b/resolve/main/sapiens_1b_epoch_173_clean.pth?download=true
```
Then I cropped this MAE example image from your paper, and run the mae demo by

```
python demo/mae_demo.py configs/sapiens_mae/humans_300m_test/mae_sapiens_1b-p16_8xb512-coslr-1600e_humans_300m_test.py ../checkpoints/pretrain/checkpoints/sapiens_1b/sapiens_1b_epoch_173.pth  --input ../data/ --output-root ../results 
```
But the results do not seem to inpaint anything meaningful for the masked tokens and tends to be gray. This seems different from what's shown in the paper. Could you help? Did I miss anything?

![paper](https://github.com/user-attachments/assets/5806dc30-5582-4c12-884e-31526b5a00cc)

**Comment by rawalkhirodkar** [comment]
@roxanneluo Hello, there are two things here:

- Ensure model weights are loaded correctly: Did you rename the _clean.pth as .pth? We remove the .backbone prefix from our model dict and to create the _clean.pth checkpoint. The mae_vis file you are using expects the prefix in the model keys.

Concretely, you will have to invert this process.

```
from mmengine.runner.checkpoint import _load_checkpoint_with_prefix
def clean_checkpoint(checkpoint_path):
    print("\033[96m" + f"Cleaning pretrained checkpoint: {checkpoint_path}" + "\033[0m")

    checkpoint = _load_checkpoint_with_prefix('backbone.', checkpoint_path)
    new_checkpoint_path = os.path.join(os.path.dirname(checkpoint_path), os.path.basename(checkpoint_path).replace('.pth', '_clean.pth'))

    print('\033[93m' + f'Saving after removing "backbone." prefix checkpoint to {new_checkpoint_path}' + '\033[0m')
    print()
    torch.save(checkpoint, new_checkpoint_path)
    return
```

- Normalized patches: The released checkpoints are trained with MAE on normalized image patches - they perform empirically better. Paper uses the unnormalized checkpoint for vis. More on this here: [](https://github.com/facebookresearch/mae/issues/12#issuecomment-1011689674)

**Comment by roxanneluo** [comment]
hi I tried to invert the process with the following code but still got the following error and bad results.

```
from argparse import ArgumentParser
from collections.abc import Sequence

from mmengine.runner.checkpoint import _load_checkpoint_with_prefix, CheckpointLoader

import torch
import os

def add_prefix_to_checkpoint(checkpoint_path):
    print("\033[96m" + f"Adding 'backbone.' prefix to checkpoint: {checkpoint_path}" + "\033[0m")

    checkpoint = _load_checkpoint(checkpoint_path)
    new_checkpoint_path = os.path.join(
        os.path.dirname(checkpoint_path),
        os.path.basename(checkpoint_path).replace('.pth', '_prefixed.pth')
    )

    print('\033[93m' + f'Saving checkpoint with added "backbone." prefix to {new_checkpoint_path}' + '\033[0m')
    print()
    torch.save(checkpoint, new_checkpoint_path)
    return

def _add_prefix_to_state_dict(prefix, state_dict):
    """Add a prefix to the keys of a state_dict.

    Args:
        prefix (str): The prefix to add.
        state_dict (dict): The state_dict whose keys will be prefixed.

    Returns:
        dict: New state_dict with prefixed keys.
    """
    if not prefix.endswith('.'):
        prefix += '.'
    prefixed_state_dict = {prefix + k: v for k, v in state_dict.items()}
    return prefixed_state_dict

def _load_checkpoint(filename, map_location='cpu', logger=None):
    """Load checkpoint from a file.

    Args:
        filename (str): Path to the checkpoint file.
        map_location (str): Device mapping. Defaults to 'cpu'.

    Returns:
        dict: The loaded checkpoint.
    """
    checkpoint = CheckpointLoader.load_checkpoint(filename, map_location, logger)
    #checkpoint = torch.load(filename, map_location=map_location)
    if 'state_dict' in checkpoint:
        checkpoint['state_dict'] = _add_prefix_to_state_dict('backbone.', checkpoint['state_dict'])
    else:
        checkpoint = _add_prefix_to_state_dict('backbone.', checkpoint)
    return checkpoint


def clean_checkpoint(checkpoint_path):
    print("\033[96m" + f"Cleaning pretrained checkpoint: {checkpoint_path}" + "\033[0m")

    checkpoint = _load_checkpoint_with_prefix('backbone.', checkpoint_path)
    new_checkpoint_path = os.path.join(os.path.dirname(checkpoint_path), os.path.basename(checkpoint_path).replace('.pth', '_clean.pth'))

    print('\033[93m' + f'Saving after removing "backbone." prefix checkpoint to {new_checkpoint_path}' + '\033[0m')
    print()
    #torch.save(checkpoint, new_checkpoint_path)
    return


def main() -> None:
    parser = ArgumentParser()
    parser.add_argument('checkpoint', help='Checkpoint file for pose')
    args = parser.parse_args()

    add_prefix_to_checkpoint(args.checkpoint)


if __name__ == "__main__":
  main()
     ```

```
/usr/local/google/home/xuluo/repos/sapiens/engine/mmengine/optim/optimizer/zero_optimizer.py:16: DeprecationWarning: `TorchScript` support for functional optimizers is deprecated and will be removed in a future PyTorch release. Consider using the `torch.compile` optimizer instead.
  from torch.distributed.optim import \
Loads checkpoint by local backend from path: ../checkpoints/pretrain/checkpoints/sapiens_1b/sapiens_1b_epoch_173.pth on server: xuluo9033953
/usr/local/google/home/xuluo/repos/sapiens/engine/mmengine/runner/checkpoint.py:354: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arb
itrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will
 no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for an
y issues related to this experimental feature.
  checkpoint = torch.load(filename, map_location=map_location)
Done: Loaded checkpoint from ../checkpoints/pretrain/checkpoints/sapiens_1b/sapiens_1b_epoch_173.pth on server: xuluo9033953
The model and loaded state dict do not match exactly   

missing keys in source state_dict: neck.mask_token, neck.decoder_pos_embed, neck.decoder_embed.weight, neck.decoder_embed.bias, neck.decoder_blocks.0.ln1.weight, neck.decoder_blocks.0.ln1.bias, neck.decoder_blocks.0.attn.qkv.weight, neck.decoder_blocks.0.attn.qkv.bias, neck.decoder_blocks.0.attn.proj.w
eight, neck.decoder_blocks.0.attn.proj.bias, neck.decoder_blocks.0.ln2.weight, neck.decoder_blocks.0.ln2.bias, neck.decoder_blocks.0.ffn.layers.0.0.weight, neck.decoder_blocks.0.ffn.layers.0.0.bias, neck.decoder_blocks.0.ffn.layers.1.weight, neck.decoder_blocks.0.ffn.layers.1.bias, neck.decoder_blocks.
1.ln1.weight, neck.decoder_blocks.1.ln1.bias, neck.decoder_blocks.1.attn.qkv.weight, neck.decoder_blocks.1.attn.qkv.bias, neck.decoder_blocks.1.attn.proj.weight, neck.decoder_blocks.1.attn.proj.bias, neck.decoder_blocks.1.ln2.weight, neck.decoder_blocks.1.ln2.bias, neck.decoder_blocks.1.ffn.layers.0.0.
weight, neck.decoder_blocks.1.ffn.layers.0.0.bias, neck.decoder_blocks.1.ffn.layers.1.weight, neck.decoder_blocks.1.ffn.layers.1.bias, neck.decoder_blocks.2.ln1.weight, neck.decoder_blocks.2.ln1.bias, neck.decoder_blocks.2.attn.qkv.weight, neck.decoder_blocks.2.attn.qkv.bias, neck.decoder_blocks.2.attn
.proj.weight, neck.decoder_blocks.2.attn.proj.bias, neck.decoder_blocks.2.ln2.weight, neck.decoder_blocks.2.ln2.bias, neck.decoder_blocks.2.ffn.layers.0.0.weight, neck.decoder_blocks.2.ffn.layers.0.0.bias, neck.decoder_blocks.2.ffn.layers.1.weight, neck.decoder_blocks.2.ffn.layers.1.bias, neck.decoder_
blocks.3.ln1.weight, neck.decoder_blocks.3.ln1.bias, neck.decoder_blocks.3.attn.qkv.weight, neck.decoder_blocks.3.attn.qkv.bias, neck.decoder_blocks.3.attn.proj.weight, neck.decoder_blocks.3.attn.proj.bias, neck.decoder_blocks.3.ln2.weight, neck.decoder_blocks.3.ln2.bias, neck.decoder_blocks.3.ffn.laye
rs.0.0.weight, neck.decoder_blocks.3.ffn.layers.0.0.bias, neck.decoder_blocks.3.ffn.layers.1.weight, neck.decoder_blocks.3.ffn.layers.1.bias, neck.decoder_blocks.4.ln1.weight, neck.decoder_blocks.4.ln1.bias, neck.decoder_blocks.4.attn.qkv.weight, neck.decoder_blocks.4.attn.qkv.bias, neck.decoder_blocks
.4.attn.proj.weight, neck.decoder_blocks.4.attn.proj.bias, neck.decoder_blocks.4.ln2.weight, neck.decoder_blocks.4.ln2.bias, neck.decoder_blocks.4.ffn.layers.0.0.weight, neck.decoder_blocks.4.ffn.layers.0.0.bias, neck.decoder_blocks.4.ffn.layers.1.weight, neck.decoder_blocks.4.ffn.layers.1.bias, neck.d
ecoder_blocks.5.ln1.weight, neck.decoder_blocks.5.ln1.bias, neck.decoder_blocks.5.attn.qkv.weight, neck.decoder_blocks.5.attn.qkv.bias, neck.decoder_blocks.5.attn.proj.weight, neck.decoder_blocks.5.attn.proj.bias, neck.decoder_blocks.5.ln2.weight, neck.decoder_blocks.5.ln2.bias, neck.decoder_blocks.5.f
fn.layers.0.0.weight, neck.decoder_blocks.5.ffn.layers.0.0.bias, neck.decoder_blocks.5.ffn.layers.1.weight, neck.decoder_blocks.5.ffn.layers.1.bias, neck.decoder_blocks.6.ln1.weight, neck.decoder_blocks.6.ln1.bias, neck.decoder_blocks.6.attn.qkv.weight, neck.decoder_blocks.6.attn.qkv.bias, neck.decoder
_blocks.6.attn.proj.weight, neck.decoder_blocks.6.attn.proj.bias, neck.decoder_blocks.6.ln2.weight, neck.decoder_blocks.6.ln2.bias, neck.decoder_blocks.6.ffn.layers.0.0.weight, neck.decoder_blocks.6.ffn.layers.0.0.bias, neck.decoder_blocks.6.ffn.layers.1.weight, neck.decoder_blocks.6.ffn.layers.1.bias,
 neck.decoder_blocks.7.ln1.weight, neck.decoder_blocks.7.ln1.bias, neck.decoder_blocks.7.attn.qkv.weight, neck.decoder_blocks.7.attn.qkv.bias, neck.decoder_blocks.7.attn.proj.weight, neck.decoder_blocks.7.attn.proj.bias, neck.decoder_blocks.7.ln2.weight, neck.decoder_blocks.7.ln2.bias, neck.decoder_blo
cks.7.ffn.layers.0.0.weight, neck.decoder_blocks.7.ffn.layers.0.0.bias, neck.decoder_blocks.7.ffn.layers.1.weight, neck.decoder_blocks.7.ffn.layers.1.bias, neck.ln1.weight, neck.ln1.bias, neck.decoder_pred.weight, neck.decoder_pred.bias

```

**Comment by roxanneluo** [comment]
Is there a checkpoint still containing the decoder blocks?

**Comment by rawalkhirodkar** [comment]
@roxanneluo please checkout the 1B encoder + decoder weights: [here](https://huggingface.co/rawalkhirodkar/sapiens-pretrain-1b-mae-decoder/tree/main).
